File: p14ab.hs
-- Advent of Code 2023 - Day 14 part One and Two
-- Solutions in Haskell
-- (Ter leering ende vermaeck...)
--
-- The total load on the north support beams: 109596
-- The total load after 1000000000 cycles:     96105
--
-- (cl) by Arno Jacobs, 2023-12-14

module AoC2023d14ab where

import Data.List

cFreeSpace      = '.' :: Char
cRollingStone   = 'O' :: Char
cSteadyRock     = '#' :: Char

megaCycles      = 1000000000 :: Int

-- Some initials
filename :: String
filename = "data/inputDay14_2023.txt"


-- Part 1

rollLeft :: String -> String
rollLeft = concat . map rollPartToLeft . seperateOn cSteadyRock

rollRight :: String -> String
rollRight = reverse . rollLeft . reverse

rollPartToLeft :: String -> String
rollPartToLeft xs   
    | elem cSteadyRock xs   = xs
    | otherwise             = theRollingStones ++ theFreeSpace
        where
            theRollingStones = filter (== cRollingStone) xs
            theFreeSpace     = filter (== cFreeSpace) xs

-- 
seperateOn :: Eq a => a -> [a] -> [[a]]
seperateOn _   []       = []
seperateOn sep xs 
    | head xs == sep    = [seperators] ++ seperateOn sep afterSeperators
    | otherwise         = [rest]       ++ seperateOn sep afterRest
        where
            seperators      = takeWhile (== sep) xs
            afterSeperators = dropWhile (== sep) xs 
            rest            = takeWhile (/= sep) xs
            afterRest       = dropWhile (/= sep) xs

-- Calculate the load of the slided rocks - line by line 
calculateLoad :: String -> Int
calculateLoad xs = sum . map fst . filter ((== cRollingStone) . snd) $ zip [1..] xs 
    
-- Transpose the dish data - so the rolling stones will roll to the "west" (left)
-- Flip dish by reverse for easy indexing and calculating load
totalLoad :: [String] -> Int
totalLoad = sum . map (calculateLoad . reverse . rollLeft) . transpose


-- Part 2

-- One cycle: North -> West -> South -> East
--
-- North:  T                 roll                 T
-- West:                     roll
-- South:  T  map (reverse)  roll  map (reverse)  T
-- East:      map (reverse)  roll  map (reverse)

rollNorth :: [String] -> [String]
rollNorth = transpose . map rollLeft . transpose

rollWest :: [String] -> [String]
rollWest = map rollLeft

rollSouth :: [String] -> [String]
rollSouth = transpose . map rollRight . transpose

rollEast :: [String] -> [String]
rollEast = map rollRight

oneCycle :: [String] -> [String]
oneCycle = rollEast . rollSouth . rollWest . rollNorth

nCycles :: Int -> [String] -> [String]
nCycles count dish  | count < 2 = oneCycle dish 
                    | otherwise = nCycles (count - 1) (oneCycle dish)

getModuloCycle :: [String] -> [String] -> Int
getModuloCycle firstCycle currentCycle 
    | nextCycle == firstCycle   = 1
    | otherwise                 = 1 + getModuloCycle firstCycle nextCycle 
        where
            nextCycle = oneCycle currentCycle

totalLoadAfterCycle :: [String] -> Int
totalLoadAfterCycle = sum . map (calculateLoad . reverse) . transpose

workToPattern :: [String] -> [[String]] -> Int -> (Int,[String])
workToPattern dish previousDishes cycles    
    | elem dish previousDishes  = (cycles,dish)
    | otherwise                 = workToPattern nextDish currentDishes (cycles+1)
        where
            nextDish        = oneCycle dish
            currentDishes   = [dish] ++ previousDishes

workMegaCycles ::  Int -> [String] -> Int
workMegaCycles cycles dish = totalLoadAfterCycle $ nCycles remainingCycles dish' 
    where
        (startCycles,dish') = workToPattern dish [] 0
        moduloCycles        = getModuloCycle dish' dish'
        remainingCycles     = mod (cycles - startCycles) moduloCycles

main :: IO ()
main = do   putStrLn "Advent of Code 2023 - day 14  (Haskell)"
            dish <- lines <$> readFile filename            

            putStr "The total load on the north support beams: "
            print $ totalLoad dish
            putStr "The total load after "
            putStr $ show megaCycles
            putStr " cycles:     "
            print $ workMegaCycles megaCycles dish

            putStrLn "0K.\n"

File: p14ab.clj
;;;;    Advent of Code 2023 - Day 14 part One and Two
;;;;    https://adventofcode.com/2023/day/14
;;;;
;;;;    Solutions in Clojure
;;;;    (Ter leering ende vermaeck...)
;;;;
;;;;    The total load on the north support beams: 109596
;;;;    The total load after 1000000000 cycles:     96105
;;;;
;;;;    (cl) by Arno Jacobs, 2023-12-14
;;;;    

(ns p14ab
  (:require [clojure.string :as str]))

;;; Read the data and split lines
(defn get-lines 
  [file]
  (->> file
       slurp
       (str/split-lines)))

;;; All characters into a 2D-list structure
(def data-set 
  (map seq (get-lines "./data/inputDay14_2023.txt")))

;;; Some initials
(def RollingStone (char \O))
(def SteadyRock   (char \#))
(def FreeSpace    (char \.))
(def mega-cycles 1000000000)

;;;  Part 1

;; Create a list of indexes for all Rolling Stones, first position has index 1
(defn RollingStones-indexes 
  [ix mirror-line] 
  (if (empty? mirror-line)
    nil
    (if (= RollingStone (first mirror-line))
      (cons ix (RollingStones-indexes (+ 1 ix) (rest mirror-line)))
      (RollingStones-indexes (+ 1 ix) (rest mirror-line)))))

;;; Get the indexes of the Rolling Stones and sum them all
(defn calculate-load 
  [mirror-line]
  (->> mirror-line
       (RollingStones-indexes 1)
       (reduce +)))

;;; Transpose a 2D-list structure
(defn transpose 
  [mx-2D]
  (apply mapv vector mx-2D))

;;; Create a list of lists, seperated on the first given element
(defn seperate-on
  [element elements]
  (if (empty? elements)
    nil
    (if (= element (first elements))
      (cons (take-while #(= element %1) elements)
            (seperate-on element (drop-while #(= element %1) elements)))
      (cons (take-while #(not= element %1) elements)
             (seperate-on element (drop-while #(not= element %1) elements))))))

;;; A mirror-line-part with a Rolling Stone will place those Stones most left
(defn roll-part-to-the-left 
  [mirror-line-part]
  (cons (filter #(= RollingStone %1) mirror-line-part) 
        (filter #(not= RollingStone %1) mirror-line-part)))

;;; Roll all rolling stones to the left, a steady rock will stop the rolling ones.
(defn roll-left
  [mirror-line]
  (->> mirror-line
       (seperate-on SteadyRock)
       (map roll-part-to-the-left)
       flatten))

;;; Handy for part 2
;;; Roll all rolling stones to the right, a steady rock will stop the rolling ones.
(defn roll-right
  [mirror-line]
  (->> mirror-line 
       reverse 
       roll-left
       reverse))

;;; Per mirror line, roll all balls left, reverse line and calculate load
(defn roll-and-calculate-load 
  [mirror-line]
  (->> mirror-line
       roll-left
       reverse
       calculate-load))

;;; Work the complete mirror, turn North anti-clockwise
(defn total-load 
  [mirror]
  (->> mirror
       transpose
       (map roll-and-calculate-load)
       (reduce +)))

;;; Part 2 

;;;; One cycle: North -> West -> South -> East
;;;;
;;;; North:  T                 roll                 T
;;;; West:                     roll
;;;; South:  T  map (reverse)  roll  map (reverse)  T
;;;; East:      map (reverse)  roll  map (reverse)

;;; Roll all rolling stones up nort
(defn roll-north 
  [mirror]
  (->> mirror
       transpose
       (map roll-left)
       transpose))

;;; Roll all rolling stones left to the west
(defn roll-west
  [mirror]
  (map roll-left mirror))

;;; Roll all rolling stones down south
(defn roll-south
  [mirror]
  (->> mirror
       transpose
       (map roll-right) 
       transpose))

;;; Roll all rolling stones right to the east
(defn roll-east
  [mirror]
  (map roll-right mirror))

;;; Roll all four directions in an anti clockwise sequence - once
(defn one-cycle 
  [mirror]
  (->> mirror 
       roll-north
       roll-west
       roll-south
       roll-east))

;;; Work a complete cycle n-times
(defn n-cycles 
  [count mirror]
  (if (< count 2)
    (one-cycle mirror)
    (n-cycles (- count 1) (one-cycle mirror))))

;;; There is repetition after some cycles. Find that cycle
(defn get-modulo-cycle
  [start-cycle current-cycle modulo-counter]
  (let [next-cycle (one-cycle current-cycle)]
    (if (= next-cycle start-cycle) 
      modulo-counter
      (get-modulo-cycle start-cycle next-cycle (+ 1 modulo-counter)))))

;;; Calculate the total load after running all cycles 
(defn total-load-after-all-cycles 
  [mirror]
  (->> mirror
       transpose
       (map reverse)
       (map calculate-load)
       (reduce +)))

;;; Calculate the total load after running all cycles  
(defn work-to-pattern-start
  [mirror previous-mirrors cycles]
  (let [in-set ((set previous-mirrors) mirror)]
    (if (nil? in-set)
      (work-to-pattern-start (one-cycle mirror) 
                             (cons mirror previous-mirrors) 
                             (+ cycles 1))
      (list cycles mirror))))

;;; The works. Find pattern start. Find cycle. Calculate remaining cycles. And work.
(defn work-mega-cycles
  [cycles mirror]
  (let [pattern-info (work-to-pattern-start mirror nil 0),
        current-mirror (first (rest pattern-info)),
        modulo-cycle (get-modulo-cycle current-mirror current-mirror 1),
        remaining-cycles (mod (- cycles (first pattern-info)) modulo-cycle)]
    (total-load-after-all-cycles (n-cycles remaining-cycles current-mirror))))

;;; The 'main' program - - -
(defn program   []
  (println "Advent of Code 2023 - day 14 (Clojure)")
  (print   "The total load on the north support beams: ")
  (println (total-load data-set))
  (print   "The total load after ")
  (print mega-cycles)
  (print " cycles:     ")
  (println (work-mega-cycles mega-cycles data-set))
  (println "0K.\n"))

;; Run in terminal via: clojure -M p14ab.clj
(program)



